torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,192 @@
+            initial_sync_ = true;
+            last_op_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
+      // Determine if any smem TV is written to at beginning of the for-loop
+      // and whether that smem TV is read from at the end of the for-loop
+      std::vector<TensorView*> common_data;
+      std::set_intersection(
+          initial_.begin(),
+          initial_.end(),
+          final_.begin(),
+          final_.end(),
+          std::back_inserter(common_data),
+          [](TensorView* left, TensorView* right) {
+            return left->name() == right->name();
+          });
+
+      // Insert new SyncThreads at end of for-loop to prevent WAR race condition
+      if (common_data.size() > 0 &&
+          fl->body().exprs().back()->getExprType().value() != ExprType::Sync &&
+          !last_op_sync_) {
+        // std::cout << "WAR race detected; Add Sync" << std::endl;
+        war_sync_ = true;
+        fl->body().push_back(new kir::Sync());
+      }
+    }
+  }
+
+  void hasOutputSmemExpr(Expr* expr, std::unordered_set<TensorView*>& set) {
+    for (auto out : expr->outputs()) {
+      if (ir_utils::isTV(out)) {
+        auto tv = out->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+  void hasInputSmemExpr(Expr* expr, std::unordered_set<TensorView*>& set) {
+    for (auto inp : expr->inputs()) {
+      if (ir_utils::isTV(inp)) {
+        auto tv = inp->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+ private:
+  explicit LocalSyncInserter(Expr* expr) {
+    handle(expr);
+  }
+
+  // Track Shared Memory Inputs (Reads) for parent for-loop
+  std::unordered_set<TensorView*> all_smem_inputs_;
```

Comment:
if the TensorViews are used only as identifiers, can we make them `const TensorView*` ? (same below)